src/main.rs

```diff
@@ -73,52 +77,53 @@ fn three_letter_code(aa: char) -> String {
     ]
     .iter()
     .copied()
-    .collect();
+    .collect()
+});
 
+/// get the three-letter equivalent of an amino acid
+/// aa (char): the amino acid to translate, eg 'A' -> "Ala"
+/// returns: String
+fn three_letter_code(aa: char) -> Result<String, Error> {
     // check input.
-    let re = Regex::new(r"[A-Z*]").unwrap();
-    if !re.is_match(&aa.to_string()) {
-        println!("Invalid amino acid: {}", aa);
-        process::exit(1);
+    match aa {
+        'A'..='Z' => Ok(THREE_LETTER_CODE[&aa].to_string()),
+        _ => Err(Error::InvalidAminoAcid(aa)),
     }
-    three_letter_map[&aa].to_string()
 }
 
-// translate a codon into its corresponding amino acid
-// triplet (&str): a three-letter codon eg "ATG"
-// i (usize): codon position. if 0, use the STARTS table
-// returns: char
-fn translate(triplet: &str, i: usize) -> char {
+/// translate a codon into its corresponding amino acid
+/// triplet (&str): a three-letter codon eg "ATG"
+/// i (usize): codon position. if 0, use the STARTS table
+/// returns: char
+fn translate(triplet: &[u8], i: usize) -> Result<char, Error> {
     let mut codon = vec![ERR_BAD_NT; 3];
 
-    for (i, base) in triplet.chars().enumerate() {
-        codon[i] = lookup(base);
+    for (i, base) in triplet.iter().enumerate() {
+        codon[i] = lookup(*base);
     }
 
     if codon.contains(&ERR_BAD_NT) {
-        println!("Something went wrong in translation!");
-        println!("{:?}", codon);
-        process::exit(1);
+        return Err(Error::BadNucleotide(codon));
     }
 
     let index: usize = (codon[0] * 16) + (codon[1] * 4) + codon[2];
     // translate the codon into single-letter code
 
-    let c = if (i == 0) && (STARTS.chars().nth(index).unwrap() == 'M') {
```

Comment:
Because `&[u8]` iterates by bytes, there is no need to call `.chars().nth()` here. This is an example of how UTF-8 makes it harder to process text (because it can represent a codepoint with more than one byte), but it is essential for correctness when working with anything out of the ASCII range.

(a very detailed and long post about this: https://fasterthanli.me/articles/working-with-strings-in-rust)